Query access API

bojanz opened the pull request:
Continued from #50.

- Replaced the old Condition class with the new ConditionGroup / Condition pair, which is cleaner, and performs more validation (the available operators, for example). We previously aimed for the class to be consistent with the core Condition classes, but it's not worth the DX hit, especially since access conditions really are their own thing.
- Refactored query access handler logic. My aim was to make their structure similar to the access handlers themselves (the split into owner conditions and non-owner conditions, for example). As a result of that, a lot more logic is reused. The status condition is now applied in both handlers, not just the cacheable one, for example.
- Added unit tests for Condition/ConditionGroup, kernel tests for the query access handlers.

TODO:
1) ~~Add back code for "view own unpublished $entity_type", it got removed in the refactor.~~
2) ~~Expand query access handler tests to cover entity types without an owner (the one provided by entity_module_test, for example).~~
3) ~~Convert SqlQueryAlter into a service and give it an interface, to allow it to be swapped.~~ (Skipped for now, staying in sync with core).
4) ~~General cleanup of SqlQueryAlter, handling of the todos, tests.~~
5) ~~An event that's fired after calling the query access handler, with access to the conditions, operation, account, allowing modules to modify the conditions before they are applied.~~

Questions:
1) ~~Should the namespace be Drupal\entity\QueryAccess instead of Drupal\entity\Query?~~ (Decided: Yes)
2) ~~Should Condition and ConditionGroup be AccessCondition and AccessConditionGroup? Not convinced right now. Would also mean that QueryAccessHandler::buildConditions() becomes buildAccessConditions()?~~ (Decided: No)
3) Is it odd that UncacheableQueryAccessHandler adds no logic to the base class? Should QueryAccessHandler extend UncacheableQueryAccessHandler instead, losing the base class completely?
4) ~~Do we gain anything from supporting update/delete operations (and not just view)? I know we've gone back & forth on that topic a few times.~~ (Decision: Since the code is already there, we might as well).

dawehner committed:
Initial work on query access.

bojanz committed:
More work on query access.

- Introduce new Condition and ConditionGroup value objects, add test coverage.
- Rewrite the query access handlers, taking into account:
  - Per bundle view permissions provided by EntityPermissionProvider.
  - Changed view permission names in UncacheablePermissionProvider.
  - The need to restrict entities to published ones in both handlers.

gabesullice commented on src/Query/Condition.php:
```diff
@@ -0,0 +1,113 @@
+/**
+ * Represents a single access condition.
+ */
+final class Condition {
```
A quick search of Drupal core shows that the `final` keyword is only used for Event objects. Although, I can see why it would be useful for this value object.

We have this same class, almost exactly for JSON API and for the work done for [Issue #777578](https://www.drupal.org/project/drupal/issues/777578), [here](https://cgit.drupalcode.org/entity_access_policies/tree/src/Condition/Value?h=8.x-1.x). It definitely seems like this is a needed concept.

bojanz replied:
I tend to stick "final" to every class whose usage is hardcoded. Conditions can't be swapped out, so there's no point in allowing them to be extended.

That said, I'm not married to this detail. If it doesn't pass core review, we'll remove it at that point.

dawehner replied:
For me sounds like a good idea to make this final. If there comes up a usecase removing it is always possible

dawehner replied:
One reason final isn't that much a thing in core is that you can't mock these classes anymore. On the other hand, Condition for example is a clear value object.

gabesullice commented on entity.module:
```diff
@@ -86,3 +89,54 @@ function entity_module_preuninstall($module) {
+  $entity_type_id = $query->getMetaData('entity_type');
+  if ($query->hasTag($entity_type_id . '_access')) {
+    $entity_type_manager = \Drupal::entityTypeManager();
+    $entity_type = $entity_type_manager->getDefinition($entity_type_id);
```
_Technically_... this could be `NULL` based on the EntityTypeManagerInterface.

dawehner replied:
Wouldn't this just throw an exception?

gabesullice replied:
Yeah, I guess that's really what we want here. I can't think of anything more appropriate.

gabesullice commented on entity.module:
```diff
@@ -86,3 +89,54 @@ function entity_module_preuninstall($module) {
+      $entity_type->getDataTable(),
+      $entity_type->getRevisionTable(),
+      $entity_type->getRevisionDataTable(),
+    ];
```
Building this map could be statically cached and moved to a helper method/function like `getEntityTypeTableNames()`, I'm not sure if that'd be worth it.

gabesullice commented on src/Query/Condition.php:
```diff
@@ -0,0 +1,113 @@
+   */
+  protected static $supportedOperators = [
+    '=', '<>', '<', '<=', '>', '>=', 'BETWEEN', 'NOT BETWEEN',
+    'IN', 'NOT IN', 'IS NULL', 'IS NOT NULL',
```
What's the reason this doesn't list `STARTS_WITH`, `CONTAINS` or `ENDS_WITH`?

bojanz replied:
We used only the operators that can be directly passed to the matching SQL condition, because of pure laziness. No reason not to extend it in a followup.

gabesullice commented on src/Query/Condition.php:
```diff
@@ -0,0 +1,113 @@
+  public function __construct($field, $value, $operator = NULL) {
+    // Provide a default based on the data type of the value.
+    if (!isset($operator)) {
+      $operator = is_array($value) ? 'IN' : '=';
```
👌

GoZOo replied:
every root functions such as is_array, in_array, count etc should be prepended by \ to improve performances. See https://veewee.github.io/blog/optimizing-php-performance-by-fq-function-calls/

dawehner replied:
I'm wondering whether this is actually the right thing to do. After drupalgeddon Nr. 1 there was this drupal core issue: https://www.drupal.org/node/2358079 which would probably argue this automatic conversion should not be done.

bojanz replied:
Definitely not prepending \ until core starts doing that. No D8 code I've seen/written (Commerce included) does that.

gabesullice commented on src/Query/Condition.php:
```diff
@@ -0,0 +1,113 @@
+   * @return string
+   *   The string representation of the condition.
+   */
+  public function __toString() {
```
❤️

gabesullice commented on src/Query/ConditionGroup.php:
```diff
@@ -0,0 +1,194 @@
+    if ($field instanceof ConditionGroup) {
+      if ($field->count() === 1) {
+        // The condition group only has a single condition, merge it.
+        $this->conditions[] = reset($field->getConditions());
```
This is really elegant.

gabesullice commented on src/Query/ConditionGroup.php:
```diff
@@ -0,0 +1,194 @@
+        $this->conditions[] = reset($field->getConditions());
+        $this->addCacheTags($field->getCacheTags());
+        $this->addCacheContexts($field->getCacheContexts());
+        $this->mergeCacheMaxAge($field->getCacheMaxAge());
```
There's a method for this: `$this->addCacheableDependency($field);`?

gabesullice replied:
Actually, I'm pretty sure this just isn't needed at all because it's always going to be computed below. This is just doing extra work.

gabesullice commented on src/Query/ConditionGroup.php:
```diff
@@ -0,0 +1,194 @@
+  }
+
+  /**
+   * Gets the string representation of the condition group.
```
Is this for development purposes also?

gabesullice replied:
It must be, because: `return (string) reset($this->conditions);`

So, maybe add that remark to this comment too?

gabesullice commented on src/Query/ConditionGroup.php:
```diff
@@ -0,0 +1,194 @@
+        $tags = array_merge($tags, $condition->getCacheTags());
+      }
+    }
+    return Cache::mergeTags($tags, []);
```
This must be evaluated for each child every time this method is called because `public function &getConditions()` would permit any pre-computed value to be invalidated 👍

gabesullice commented on src/Query/QueryAccessHandler.php:
```diff
@@ -0,0 +1,28 @@
+  /**
+   * {@inheritdoc}
+   */
+  protected function buildEntityOwnerConditions($operation, AccountInterface $account) {
```
This seems like a very specific method name and purpose for such a generic class name. Is this meant to be an example implementation?

bojanz replied:
The method names were made to match the names in EntityAccessControlHandlerBase (and the child EntityAccessControlHandler / UncacheableAccessControlHandler). Since the query handlers match the access control handler logic, it felt right to have their general flow match as well.

gabesullice commented on src/Query/QueryAccessHandlerBase.php:
```diff
@@ -0,0 +1,210 @@
+   * {@inheritdoc}
+   */
+  public function buildConditions($operation, AccountInterface $account = NULL) {
+    $account = $account ?: $this->currentUser;
```
👍

gabesullice commented on src/Query/QueryAccessHandlerBase.php:
```diff
@@ -0,0 +1,210 @@
+    $account = $account ?: $this->currentUser;
+    $entity_type_id = $this->entityType->id();
+
+    if ($account->hasPermission("administer {$entity_type_id}")) {
```
This won't work with `node` will it? The corresponding permission is `administer content types`. Perhaps this is fixed by your permission handler patch.

bojanz replied:
No, these handlers are not meant to be used for node (or any entity type that doesn't use our permission handlers). I'd expect node to get a custom query_access handler, until we can deprecate the old permissions and switch node to one of the permission handlers.

gabesullice replied:
👍

gabesullice commented on src/Query/QueryAccessHandlerBase.php:
```diff
@@ -0,0 +1,210 @@
+    }
+
+    if ($this->entityType->entityClassImplements(EntityOwnerInterface::class)) {
+      $entity_conditions = $this->buildEntityOwnerConditions($operation, $account);
```
I wonder if we could `assert` somewhere that the query access handler for Foo entity type should have define this method if the interface is implemented. That way, one would figure out that it needs to be implemented while creating the entity type, rather than the first time an entity query is executed.

bojanz replied:
The method is defined in the same class, it can't be un-defined.

gabesullice replied:
Derp, I see it below now :)

I saw it redefined above and thought that it must be implemented on whatever class extended from this one.

gabesullice reviewed (commented):
Overall, this is incredible work! I'm really excited to see an API for this sort of thing. I'll leave a more high-level summary in the Drupal.org issue queue and leave the line-by-line review here.

GoZOo commented on src/Query/Condition.php:
```diff
@@ -0,0 +1,113 @@
+   */
+  public function __construct($field, $value, $operator = NULL) {
+    // Provide a default based on the data type of the value.
+    if (!isset($operator)) {
```
Prefer using `NULL === $operator`

GoZOo commented on src/Query/QueryAccessHandler.php:
```diff
@@ -0,0 +1,28 @@
+   * {@inheritdoc}
+   */
+  protected function buildEntityOwnerConditions($operation, AccountInterface $account) {
+    if ($operation == 'view') {
```
you can safely use `===` here

GoZOo commented on src/Query/QueryAccessHandlerBase.php:
```diff
@@ -0,0 +1,210 @@
+    }
+
+    $conditions = NULL;
+    if ($operation == 'view' && $this->entityType->entityClassImplements(EntityPublishedInterface::class)) {
```
same thing here, you can use `===` instead

GoZOo commented on src/Query/SqlQueryAlter.php:
```diff
@@ -0,0 +1,172 @@
+   *   The cacheability metadata.
+   */
+  protected function applyCacheability(CacheableMetadata $cacheable_metadata) {
+    $request = \Drupal::requestStack()->getCurrentRequest();
```
Is there any reasons to not use dependency injection ?

wimleers commented on src/Query/ConditionGroup.php:
```diff
@@ -0,0 +1,194 @@
+ */
+final class ConditionGroup implements \Countable, CacheableDependencyInterface {
+
+  use RefinableCacheableDependencyTrait;
```
This implements `CacheableDependencyInterface`, so `RefinableCacheableDependencyTrait` is wrong, because it adds mutability methods. You want `CacheableDependencyTrait` instead.

bojanz replied:
We are definitely using RefinableCacheableDependencyTrait's methods (addCacheContexts), the used interface is wrong.

wimleers commented on src/Query/SqlQueryAlter.php:
```diff
@@ -0,0 +1,172 @@
+    if ($request->isMethodCacheable() && $renderer->hasRenderContext()) {
+      $build = [];
+      $cacheable_metadata->applyTo($build);
+      $renderer->render($build);
```
This bubbles onto the global render context just like we do in `\Drupal\Core\Render\MetadataBubblingUrlGenerator::bubble()`. But we did _that_ because we had to layer it on top after the fact. Ideally, you don't need this bubbling.

I suspect you need it here too because doing otherwise would break BC of the Entity Query API?

bojanz replied:
This is dawehner's code, made to follow the core example. Can't say I understand it, so I kept it.

gabesullice commented on entity.module:
```diff
@@ -86,3 +89,54 @@ function entity_module_preuninstall($module) {
+ * @return \Drupal\Core\Entity\EntityTypeInterface|null
+ *   The entity type, or NULL if unknown.
+ */
+function _entity_get_entity_type_id_from_table($table_name) {
```
The function name is "entity_type_id" but returns an entity type.

bojanz replied:
Good catch

dawehner commented on src/Query/QueryAccessHandlerBase.php:
```diff
@@ -0,0 +1,210 @@
+
+    // Own $bundle permission.
+    foreach ($bundles as $bundle) {
+      if ($account->hasPermission("$operation own $bundle $entity_type_id")) {
```
I have a hard time arguing about cache contexts, but it somehow feels like in order to support cacheability on the query level we would need to add the context outside of the permission. @wimleers can probably give this another review as well.

dawehner commented on src/Query/SqlQueryAlter.php:
```diff
@@ -0,0 +1,172 @@
+    }
+
+    // @todo Is this the right join?
+    return $select->innerJoin($table, NULL, $entity_type->getBaseTable() . '.' . $entity_type->getKey('id') . ' = ' . '%alias.entity_id');
```
Looking into ```\Drupal\Core\Entity\Query\Sql\Condition::compile``` it seems to be possible to be both

gabesullice commented on src/Query/QueryAccessHandler.php:
```diff
@@ -0,0 +1,28 @@
+    }
+    else {
+      return parent::buildEntityOwnerConditions($operation, $account);
+    }
```
No need for the `else` here.

gabesullice commented on src/Query/QueryAccessHandlerBase.php:
```diff
@@ -0,0 +1,210 @@
+    if ($account->hasPermission("administer {$entity_type_id}")) {
+      // The user has full access to all operations, no conditions needed.
+      $conditions = new ConditionGroup('OR');
+      $conditions->addCacheContexts(['user.permissions']);
```
@wimleers this is why `RefineableCacheableDependencyInterface` was needed.

@bojanz: In JSON API, we had similar value objects and dealt quite a bit with access related caching. We found it to be a major pain point and source of bugs. To help make caching more reasonable and less error-prone, @wimleers added a constructor argument to all the value objects and [got rid of the `RefineableCacheableDependency` stuff.](https://www.drupal.org/project/jsonapi/issues/2940342#comment-12517745). That essentially made everything immutable.

I recommend that cacheability be an explicit argument to the condition constructor. Given that we're dealing with access caching, I think it's acceptable, even advisable, to force implementors to think about caching, despite the slight DX loss and tediousness of it.

That means: removing `RefinableCacheableDependencyTrait`. Adding an argument to `ConditionGroup`'s constructor.

Also, why not make the Condition themselves cacheable, for most cases, they're really what are likely to change, not the group.

bojanz replied:
This was attempted in https://github.com/bojanz/entity/pull/5, but it feels like the price in terms of DX regressions is too high.

gabesullice replied:
I think two things are getting mixed together:

1. Making `CacheableMetadata` a required argument to `Condition::__construct`
2. Passing a `Condition` object to `ConditionGroup::addCondition` vs. a series of arguments and letting `addCondition` make the `new Condition()` call.

My sense is that your DX concern is actually more about `2` than `1`. Is that right?

If so, `2` was honestly just a personal preference of mine that I should have kept out of the PR.

Making all arguments to `Condition::__construct` required, cacheability included, is the real change there and I think it's a worthy price to pay for correctness.

gabesullice replied:
https://github.com/fago/entity/pull/52#discussion_r199596243 already pointed out that `$operator` should probably just be required anyway.

bojanz committed:
Cleanup.

- Address initial review feedback.
- Move buildConditions() logic to getConditions(), to prepare for a dispatched event.

bojanz committed:
Fire a QueryAccessEvent from the handlers. Introduce ConditionGroup::alwaysFalse().

bojanz committed:
Change the namespace for clarity (Drupal\entity\Query -> Drupal\entity\QueryAccess)

kristiaanvandeneynde commented on src/QueryAccess/QueryAccessHandlerBase.php:
```diff
@@ -0,0 +1,243 @@
+   */
+  public function buildConditions($operation, AccountInterface $account) {
+    $entity_type_id = $this->entityType->id();
+    if ($account->hasPermission("administer {$entity_type_id}")) {
```
Might be cleaner with a method that returns the admin permission instead of concatenating the string here. That said, I can see more of this concatenating below where using a method might not be as clean...

bojanz replied:
Yep, it's a no-go cause we'd have a dozen methods. This (concat) is also what the matching permission provider does.

kristiaanvandeneynde commented on src/QueryAccess/QueryAccessHandlerBase.php:
```diff
@@ -0,0 +1,243 @@
+          ->addCondition($bundle_key, $bundle)
+        );
+      }
+    }
```
In case we're not re-using the code from buildEntityConditions(), the above loop could be pulled into the other bundle loop, populate a  $bundles_with_own_permission array and then add that as follows:
```    // Own $bundle permission.
    if ($bundles_with_own_permission) {
        $conditions->addCacheContexts(['user']);
        $conditions->addCondition((new ConditionGroup('AND'))
          ->addCondition($uid_key, $account->id())
          ->addCondition($bundle_key, $bundles_with_own_permission)
        );
    }
```

Right now, you're adding the same condition group for every bundle.

bojanz replied:
Good point.

kristiaanvandeneynde commented on src/QueryAccess/QueryAccessHandlerBase.php:
```diff
@@ -0,0 +1,243 @@
+   * Builds the conditions for entities that have an owner.
+   *
+   * @param string $operation
+   *   The access operation. One of "view", "update" or "delete".
```
Can we not limit ourselves to these operations? It would be nice to support an "add-to-cart" operation for example.

bojanz replied:
We should follow core and say "Usually one of". The value is not actually validated anywhere.

kristiaanvandeneynde replied:
"Usually one of" would be nice.

kristiaanvandeneynde commented on src/QueryAccess/QueryAccessHandlerBase.php:
```diff
@@ -0,0 +1,243 @@
+   * @return \Drupal\entity\QueryAccess\ConditionGroup|null
+   *   The conditions, or NULL if the user doesn't have access to any entity.
+   */
+  protected function buildEntityOwnerConditions($operation, AccountInterface $account) {
```
Perhaps not copy the logic from buildEntityConditions(), but call it and build on top of its return value?

bojanz replied:
Both methods are distinctive enough that we should not attempt to reuse code between them. The logic needs to be as simple and explicit as possible, to avoid bugs.

Also note that in general, different permissions are checked, nothing is completely identical.

kristiaanvandeneynde replied:
Yeah I read over the "any" part that's only available in one of both methods.

bojanz committed:
Handle the "view own unpublished $entity_type" permission.

bojanz committed:
Cleanup (kristiaanvandeneynde's review).

bojanz committed:
Clean up tests, remove custom test module. Fix bug in QueryAccessHandlerBase.

kristiaanvandeneynde commented:
Could we have a (perhaps optional) provider set in the Condition and ConditionGroup constructors? The handler knows about this by calling `$this->entityType->getProvider()`.

The benefit would be that, if you know module A clashes with module B, module B could easily find module A's conditions in its event subscriber and do something with that. Otherwise, they'd have to inspect every Condition or ConditionGroup to try and guess which ones they want to change.

Adding the provider is just one way of doing it, though. The main idea is to be able to trace a condition to its source or to be able to figure out who added a condition for what reason.

bojanz committed:
Replace SqlQueryAlter with new EntityQueryAlter and ViewsQueryAlter classes. Rewrite and expand tests.

wimleers commented on src/QueryAccess/Condition.php:
```diff
@@ -0,0 +1,113 @@
+namespace Drupal\entity\QueryAccess;
+
+/**
+ * Represents a single access condition.
```
*Query* access condition?

wimleers commented on src/QueryAccess/Condition.php:
```diff
@@ -0,0 +1,113 @@
+  /**
+   * The supported operators.
+   *
+   * @var array
```
Nit: `string[]`

wimleers commented on src/QueryAccess/Condition.php:
```diff
@@ -0,0 +1,113 @@
+   *                   IN, NOT IN, IS NULL, IS NOT NULL.
+   */
+  public function __construct($field, $value, $operator = NULL) {
+    // Provide a default based on the data type of the value.
```
Why even provide a default? Why not require explicitness?

bojanz replied:
Because that's what core does in its own conditions (SQL / EntityQuery), and I wanted the condition API to be similar. I could make the operator required and move the default setting to ConditionGroup::addCondition() but I'm not sure that helps. And if there's no default on any level, then the operator must be specified even when it's '=', which makes it inconsistent with core.

wimleers replied:
Core also has far too many "magic defaults". But I defer to you to decide on this one — I don't feel very strongly about it. I'm just pointing out the addition of new magic. But if that magic matches what core already does, and it does so for good reason, then 👍

wimleers replied:
Hm, looking at this again, core has `\Drupal\Core\Database\Query\ConditionInterface`, which is _very_ different from the `Condition` and `ConditionGroup` classes being added here. Neither of those classes even implements core's interface. I understand why they don't. But then I don't understand why we want it to be as similar as possible if it's already so dissimilar?

bojanz replied:
Decided to punt on this for now, but if it gets raised again (core review or here) I have nothing against changing it.

wimleers commented on src/QueryAccess/Condition.php:
```diff
@@ -0,0 +1,113 @@
+  /**
+   * Gets the string representation of the condition.
+   *
+   * Used for debugging purposes.
```
👏

wimleers commented on src/QueryAccess/ConditionGroup.php:
```diff
@@ -0,0 +1,227 @@
+ *   );
+ * @endcode
+ */
+final class ConditionGroup implements \Countable, RefinableCacheableDependencyInterface {
```
Implementing `RefinableCacheableDependencyInterface` (`RCDI`) makes the cacheability of this value object mutable. I don't see why this is desirable.

wimleers replied:
Consider this resolved.

wimleers commented on src/QueryAccess/ConditionGroup.php:
```diff
@@ -0,0 +1,227 @@
+   *
+   * @return $this
+   */
+  public function addCondition($field, $value = NULL, $operator = NULL) {
```
So this is itself a mutable value object. Let's assume this is necessary. Then why do we need the cacheability of this value object to be mutable too?

bojanz replied:
Because I could not figure out a decent DX without it. Gabe has a proposed PR linked above that tried to remove mutability, but introduced a DX regression. Would probably be better to discuss on Slack.

wimleers replied:
I tried, and failed too.

The similarity between interactions with `ConditionGroup` and `AccessResult` seem reasonable. I think that makes sense. So I'm fine with keeping this as-is.

wimleers replied:
Consider this resolved.

wimleers commented on src/QueryAccess/ConditionGroup.php:
```diff
@@ -0,0 +1,227 @@
+        $this->conditions[] = reset($field->getConditions());
+        $this->addCacheTags($field->getCacheTags());
+        $this->addCacheContexts($field->getCacheContexts());
+        $this->mergeCacheMaxAge($field->getCacheMaxAge());
```
These 3 method calls are where you use `RCDI`'s capabilities. But … you could also just implement `CacheableDependencyInterface` (`CDI`) and then use `\Drupal\Core\Cache\CacheableDependencyTrait::setCacheability(CacheableMetadata::createFromObject($this)->merge($field)`.

By doing this, you keep the *value* of this value object mutable (`addCondition()` continues to work as it does today), but you keep the *cacheability* of this value object immutable. IOW: cacheability is mutated only when the value is mutated.

wimleers replied:
File uploads on GH are failing so, here is what I'm proposing:
```
diff --git a/src/QueryAccess/ConditionGroup.php b/src/QueryAccess/ConditionGroup.php
index 3dca874..f64c570 100644
--- a/src/QueryAccess/ConditionGroup.php
+++ b/src/QueryAccess/ConditionGroup.php
@@ -3,8 +3,9 @@
 namespace Drupal\entity\QueryAccess;
 
 use Drupal\Core\Cache\Cache;
-use Drupal\Core\Cache\RefinableCacheableDependencyInterface;
-use Drupal\Core\Cache\RefinableCacheableDependencyTrait;
+use Drupal\Core\Cache\CacheableDependencyInterface;
+use Drupal\Core\Cache\CacheableDependencyTrait;
+use Drupal\Core\Cache\CacheableMetadata;
 
 /**
  * Represents a group of access conditions.
@@ -33,9 +34,9 @@ use Drupal\Core\Cache\RefinableCacheableDependencyTrait;
  *   );
  * @endcode
  */
-final class ConditionGroup implements \Countable, RefinableCacheableDependencyInterface {
+final class ConditionGroup implements \Countable, CacheableDependencyInterface {
 
-  use RefinableCacheableDependencyTrait;
+  use CacheableDependencyTrait;
 
   /**
    * The conjunction.
@@ -110,9 +111,7 @@ final class ConditionGroup implements \Countable, RefinableCacheableDependencyIn
       if ($field->count() === 1) {
         // The condition group only has a single condition, merge it.
         $this->conditions[] = reset($field->getConditions());
-        $this->addCacheTags($field->getCacheTags());
-        $this->addCacheContexts($field->getCacheContexts());
-        $this->mergeCacheMaxAge($field->getCacheMaxAge());
+        $this->setCacheability(CacheableMetadata::createFromObject($this)->merge($field));
       }
       elseif ($field->count() > 1) {
         $this->conditions[] = $field;

```

wimleers commented on src/QueryAccess/ConditionGroup.php:
```diff
@@ -0,0 +1,227 @@
+      $this->conditions[] = new Condition($field, $value, $operator);
+    }
+
+    return $this;
```
Having written the above, I'm not sure it's even necessary for this `addCondition()` to mutate the current object. Why not make it a `static` method that returns a new instance? I've learned my lesson with mutable value objects … I always regret it after some time! If you feel strongly about this: sure, let's keep it this way.

But I really hope you'll consider implementing `CDI` rather than `RCDI`.

wimleers commented on src/QueryAccess/ConditionGroup.php:
```diff
@@ -0,0 +1,227 @@
+   * @return bool
+   *   Whether the condition group is always FALSE.
+   */
+  public function isAlwaysFalse() {
```
I find this name very confusing. Suggestion: `ConditionGroup::alwaysForbidden()`.

bojanz replied:
So do I, but that's what core decided to use: https://www.drupal.org/project/drupal/issues/2986334

wimleers replied:
Aha, so it's just matching core. Okay!

wimleers commented on src/QueryAccess/QueryAccessEvent.php:
```diff
@@ -0,0 +1,92 @@
+   *
+   * If $conditions->isAlwaysFalse() is FALSE, and the condition group is
+   * empty (count is 0), the user has full access, and the query doesn't
+   * need to be restricted.
```
I don't understand this comment.

wimleers commented on src/QueryAccess/QueryAccessHandlerBase.php:
```diff
@@ -0,0 +1,268 @@
+        $conditions = new ConditionGroup('OR');
+        $conditions->addCondition($published_conditions);
+        $conditions->addCondition($unpublished_conditions);
+      }
```
I just want to applaud the elegance of the code added here (in the `Handle the "view own unpublished $entity_type" permission.` commit) — it's beautiful! 😍

wimleers commented on src/QueryAccess/QueryAccessHandlerInterface.php:
```diff
@@ -0,0 +1,41 @@
+/**
+ * Query access handlers control access to entities in queries.
+ *
+ * An entity defines a query access handler in its annotation:
```
Nit: s/entity/entity type/

wimleers commented on src/QueryAccess/QueryAccessHandlerInterface.php:
```diff
@@ -0,0 +1,41 @@
+ * to filter it. For example, if the user #22 only has access to view
+ * their own entities, a uid = '22' condition will be built and applied.
+ *
+ * The following query types are supported:
```
Nit: s/query types/database query types/ (I know this is implied, but better to be explicit?)

wimleers commented on src/QueryAccess/QueryAccessHandlerInterface.php:
```diff
@@ -0,0 +1,41 @@
+  /**
+   * Gets the conditions for the given operation and user.
+   *
+   * The "entity.query_access.$entity_type_id" event is fired to allow other
```
Nit: s/other//

wimleers commented on src/QueryAccess/SqlQueryAlter.php:
```diff
@@ -0,0 +1,176 @@
+      }
+    }
+
+    // @todo Is this the right join?
```
Who would be the best person to review this?

wimleers commented on src/QueryAccess/SqlQueryAlter.php:
```diff
@@ -0,0 +1,176 @@
+        $table_alias = static::ensureTable($select, $table_name, $entity_type);
+
+        // @todo we need support for non main properties?
+        // @todo we need support for revision queries?
```
How do we proceed with these two todos?

bojanz replied:
Unfortunately, resolving these todos required a multi-day effort to rewrite the entire query access logic. Commit coming today.

wimleers replied:
> Unfortunately, resolving these todos required a multi-day effort to rewrite the entire query access logic.

😨 

> Commit coming today.

🎉

wimleers commented on src/QueryAccess/SqlQueryAlter.php:
```diff
@@ -0,0 +1,176 @@
+   * @param \Drupal\Core\Cache\CacheableMetadata $cacheable_metadata
+   *   The cacheability metadata.
+   */
+  protected function applyCacheability(CacheableMetadata $cacheable_metadata) {
```
This is meant to mimic the automatic cacheability bubbling in `node_query_node_access_alter()`, right? Let's add an `@see` to minimize the mystery here.

wimleers commented on tests/modules/entity_module_test/src/EventSubscriber/QueryAccessSubscriber.php:
```diff
@@ -0,0 +1,44 @@
+    }
+    elseif ($account->getEmail() == 'user2@example.com') {
+      // This user should have access to entities with the IDs 1, 2, and 3.
+      $conditions->alwaysFalse(FALSE);
```
Why is this `->alwaysFalse(FALSE)` call necessary?

bojanz replied:
Because the assumption is that the original query access handler has already set $query->alwaysFalse() due to insufficient permissions, so the current value is TRUE.
I can expand the comment to clarify that.

wimleers replied:
> I can expand the comment to clarify that.

That'd be great 🙏

wimleers reviewed (changes requested):
I reviewed the commits made in the past 17 days (they started landing just as I was moving across the country, bad timing on my part: otherwise I'd have been able to give feedback sooner!).

Great progress! 👍 👌

---

There is one open question in the IS:
> 3. Is it odd that UncacheableQueryAccessHandler adds no logic to the base class? Should QueryAccessHandler extend UncacheableQueryAccessHandler instead, losing the base class completely?

Yes, this is odd.

---

Review of the current PR:

- Observation: Technically this only works for entities stored and queried in a database-backed entity storage. Not the fault of this issue though. Solving this in a storage-agnostic way is vastly out of scope.
- Namespace change for clarity: +1!
- Rest of the review: left comments on the files themselves :)

wimleers commented:
Addressed @bojanz' feedback to my PR review.

bojanz commented:
The new query alter code is up. It has taken me quite a number of days to get it right, mostly due to complexities in ViewsQueryAlter (and underlying core bugs such as https://drupal.org/node/3001146).

I initially tried to use Tables for both alters (EntityQuery and Views) but that failed because of mismatched assumptions around base tables (Views will use the data table as the base table when it exists, EntityQuery doesn't do that). Discussed with dawehner and we agreed that it wasn't safe to continue pursuing.

When landing this in core we'd add an alter query to EntityQuery (which was never added to D8 because of a "lack of use cases", even though EFQ had it in D7). That would allow us to simplify EntityQueryAlter, removing all sql-dependent code from mapConditions().

Changes:
- New tests cover EntityQuery and Views, regular queries and revision queries (which use a different set of tables, and thus need tests of their own)
- EntityQueryAlter uses the EntityQuery's Tables class to handle the tough parts. Duplicates the general flow of EntityQuery's condition classes. 
- ViewsQueryAlter operates on the Views query, not the underlying SQL query. This allows for simpler logic, since we can make Views-specific assumptions, such as the data table being used as the base table when it exists.
- We now properly handle case sensitive fields (using LIKE/NOT LIKE for comparisons) in both alter classes.

Also note that I didn't introduce interfaces for the two alter classes, cause I saw that core uses this same approach, in the workspaces module (ViewsQueryAlter class).

TODO: 
- We have no test coverage for filtering by configurable field, add it.
- ~~EntityQueryAlter supports non-main properties, ViewsQueryAlter doesn't. Unify, document,~~ and add test coverage. Could also be a followup.

gabesullice commented:
@bojanz, those changes look good :) If @dawehner hasn't already done so, it'd be great if he reviewed the ViewsQueryAlter stuff.

bojanz committed:
Convert ConditionGroupTest to a kernel test, to make the build pass.

bojanz committed:
Address Wim's review.

bojanz committed:
ViewsQueryAlter should support specified property names.

bojanz committed:
Add test coverage for conditions on configurable fields, explicit property names, case sensitivity.

bojanz committed:
Fix the build.

bojanz committed:
Update issue reference.

wimleers commented:
It landed! https://www.drupal.org/project/entity/issues/2909970#comment-12794003

bojanz commented:
Thanks, everyone!
